LibWeb: (totally legit) CSS Animations

alimpfard opened the pull request:
Depends on #19037.

This partially implements CSS-Animations-1 (though there are references
to CSS-Animations-2).
Current limitations:
- Multi-selector keyframes are not supported.
- Most animation properties are ignored.
- Timing functions are not applied.
- Non-absolute values are not interpolated unless the target is also of
  the same non-absolute type (e.g. 10% -> 25%, but not 10% -> 20px).
- The JavaScript interface is left as an exercise for the next poor soul
  looking at this code.

With those said, this commit implements:
- Interpolation for most common types
- Proper keyframe resolution (including the synthetic from-keyframe
  containing the initial state)
- Properly driven animations, and proper style invalidation

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -3683,7 +3683,7 @@ ErrorOr<RefPtr<StyleValue>> Parser::parse_dimension_value(ComponentValue const&
     // 2) It's a 0.
     // We handle case 1 here. Case 2 is handled by NumericStyleValue pretending to be a LengthStyleValue if it is 0.
```
Ah, this comment says I missed this when I changed parsing. We no longer do the pretending. Apparently no layout tests happened to rely on this. 🤔

AtkinsSJ commented:
This is crashing for me when I visit `css-animations.html` in Ladybird.

```
Thread 1 "WebContent" received signal SIGSEGV, Segmentation fault.
0x00007f24a7af188b in ref () at /home/sam/serenity/Meta/Lagom/../../AK/RefCounted.h:28
28              ++m_ref_count;
(gdb) bt
#0  0x00007f24a7af188b in ref () at /home/sam/serenity/Meta/Lagom/../../AK/RefCounted.h:28
#1  NonnullRefPtr () at /home/sam/serenity/Meta/Lagom/../../AK/NonnullRefPtr.h:84
#2  copy_ () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:68
#3  copy_ () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:70
#4  copy_ () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:70
#5  Variant () at /home/sam/serenity/Meta/Lagom/../../AK/Variant.h:305
#6  Array () at /home/sam/serenity/Meta/Lagom/../../AK/Array.h:28
#7  ResolvedKeyFrame () at /home/sam/serenity/Meta/Lagom/../../Userland/Libraries/LibWeb/CSS/StyleComputer.h:137
#8  operator() () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:2043
#9  call () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:173
#10 0x00007f24a7a5f229 in operator() () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:111
#11 for_each_effective_keyframes_at_rule () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp:175
#12 0x00007f24a7af11d6 in operator()<Web::CSS::CSSStyleSheet> () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:2029
#13 0x00007f24a7aef29a in for_each_stylesheet<(lambda at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1979:41)> ()
    at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:197
#14 make_rule_cache_for_cascade_origin () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1979
#15 0x00007f24a7aeeff4 in build_rule_cache () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:2135
#16 0x00007f24a7aeece3 in build_rule_cache_if_needed () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1965
#17 compute_style_impl () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1906
#18 0x00007f24a7aeec29 in compute_style () at /home/sam/serenity/Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:1895
#19 0x00007f24a7b50f51 in recompute_style () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Element.cpp:446
#20 0x00007f24a7b36fa8 in update_style_recursively () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:915
#21 0x00007f24a7b37048 in operator()<Web::DOM::Node> () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:928
#22 for_each_child<(lambda at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:926:29)> ()
    at /home/sam/serenity/Meta/Lagom/../../Userland/Libraries/LibWeb/DOM/Node.h:535
#23 update_style_recursively () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:926
#24 0x00007f24a7b36d16 in update_style () at /home/sam/serenity/Userland/Libraries/LibWeb/DOM/Document.cpp:950
#25 0x00007f24a7dcaa20 in operator() () at /home/sam/serenity/Meta/Lagom/../../Userland/Libraries/LibJS/SafeFunction.h:81
#26 operator() () at /home/sam/serenity/Userland/Libraries/LibWeb/Platform/TimerSerenity.cpp:23
#27 call () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:173
#28 0x00007f24a5c837a6 in operator() () at /home/sam/serenity/Meta/Lagom/../../AK/Function.h:111
#29 timer_event () at /home/sam/serenity/Userland/Libraries/LibCore/Timer.cpp:78
#30 0x00007f24a5c6dca9 in dispatch_event () at /home/sam/serenity/Userland/Libraries/LibCore/Object.cpp:214
#31 0x000055d0167ff05e in qt_timer_fired () at /home/sam/serenity/Ladybird/EventLoopImplementationQt.cpp:91
#32 operator() () at /home/sam/serenity/Ladybird/EventLoopImplementationQt.cpp:106
#33 call () at /usr/include/x86_64-linux-gnu/qt6/QtCore/qobjectdefs_impl.h:146
#34 call<QtPrivate::List<>, void> () at /usr/include/x86_64-linux-gnu/qt6/QtCore/qobjectdefs_impl.h:256
#35 impl () at /usr/include/x86_64-linux-gnu/qt6/QtCore/qobjectdefs_impl.h:420
#36 0x00007f24a686a023 in ?? () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#37 0x00007f24a68784be in QTimer::timeout(QTimer::QPrivateSignal) () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#38 0x00007f24a685e62f in QObject::event(QEvent*) () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#39 0x00007f24a6811a58 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#40 0x00007f24a697a203 in QTimerInfoList::activateTimers() () from /lib/x86_64-linux-gnu/libQt6Core.so.6
#41 0x00007f24a6a39bfc in ?? () from /lib/x86_64-linux-gnu/libQt6Core.so.6
```

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:
```diff
@@ -972,6 +1273,9 @@ ErrorOr<void> StyleComputer::compute_cascaded_values(StyleProperties& style, DOM
         did_match_any_pseudo_element_rules = true;
     }
 
+    // matching_rule_set.animation_rules = collect_matching_rules(element, CascadeOrigin::Animation, pseudo_element);
+    // sort_matching_rules(matching_rule_set.animation_rules);
```
Is this meant to be a FIXME? Or did you just leave it here by mistake?

alimpfard replied:
Leftover from previous impl, will remove.

AtkinsSJ commented on Userland/Libraries/LibWeb/Forward.h:
```diff
@@ -70,10 +70,13 @@ class BorderRadiusStyleValue;
 class CalculatedStyleValue;
 class Clip;
+class CSSConditionRule;
+class CSSKeyframeRule;
+class CSSKeyframesRule;
 class ColorStyleValue;
+class CompositeStyleValue;
 class ConicGradientStyleValue;
 class ContentStyleValue;
-class CSSConditionRule;
 class CSSFontFaceRule;
```
Nit: This sorting confuses me.

alimpfard replied:
Yeah I suppose I should move all the CSS* to the top, my bad.

AtkinsSJ reviewed (changes requested):
Some stuff. My brain melted a bit at the end there so I haven't really reviewed the "actually do animationy stuff" parts.

awesomekling commented on Userland/Libraries/LibWeb/CSS/StyleComputer.h:
```diff
@@ -145,6 +164,67 @@ class StyleComputer {
+            Backward,
+            Both,
+        } fill_mode;
+        DOM::Element* owning_element;
```
Let's at least use `WeakPtr` for this :)

awesomekling commented:
The use of `Element` and `CSSStyleDeclaration` raw pointers looks sketchy.

alimpfard commented on Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp:
```diff
@@ -3163,6 +3166,117 @@ CSSRule* Parser::convert_to_rule(NonnullRefPtr<Rule> rule)
+                auto token = child_tokens.next_token();
+                if (token.is_block()) {
+                    auto block_tokens = token.block().values();
+                    block_tokens.prepend(Token::create_curly(Token::Type::OpenCurly));
+                    block_tokens.prepend(Token::create_delim("*"_short_string));
+                    block_tokens.append(Token::create_curly(Token::Type::CloseCurly));
```
It's not immediately obvious to me how this is supposed to work, I want a CSSRuleList, not a CSSStyleDeclaration?

alimpfard commented:
> The use of Element and CSSStyleDeclaration raw pointers looks sketchy

The ones in `AnimationKey`? They're just used together as a key, not dereferenced and not used otherwise.

if you mean `owning_element`, what else are we supposed to hold? JS::GCPtr is effectively an annotation type, if this is sketchy, everything else in StyleComputer is also sketchy :yakfused:

alimpfard commented:
I've replaced the one theoretically-unsafe `Element*` with a `JS::NonnullGCPtr`, and left a FIXME for removing the animation if its corresponding element isn't around.

konradekk commented on Userland/Libraries/LibWeb/CSS/CSSRuleList.cpp:
```diff
@@ -141,6 +142,38 @@ void CSSRuleList::for_each_effective_style_rule(Function<void(CSSStyleRule const
         case CSSRule::Type::Supports:
             static_cast<CSSSupportsRule const&>(*rule).for_each_effective_style_rule(callback);
             break;
+        case CSSRule::Type::Keyframe:
+        case CSSRule::Type::Keyframes:
+            break;
```
a `default` case with a debug log here too? 🤔

(also in other similar places, optionally: a `VERIFY`?)

alimpfard replied:
No. Not having a default makes the compiler tell you where you need to handle the new enum value.

konradekk replied:
neat! (missed that feature!)

konradekk commented on Userland/Libraries/LibWeb/CSS/Properties.json:
```diff
@@ -30,6 +30,103 @@
     ]
   },
+  "animation": {
+    "affects-layout": true,
```
(hmm… this depends on what exactly is animated… 🤔)

alimpfard replied:
Not exactly, changing the animation property itself will never affect layout, the animation _running_ can of course, affect layout by changing properties that do.

konradekk replied:
hah! sounds reasonable!

konradekk commented on Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:
```diff
@@ -953,6 +966,319 @@ static ErrorOr<void> cascade_custom_properties(DOM::Element& element, Optional<C
+void StyleComputer::ensure_animation_timer() const
+{
+    if (!m_animation_driver_timer) {
+        m_animation_driver_timer = Platform::Timer::create_repeating(16, [this] {
```
i remember that `16` is assuming 60fps — not necessarily obvious for other people and maybe weʼd like to control that in the future — care to extract this somewhere, at least  as a constant?

(same for `16` below)

alimpfard replied:
sure.

konradekk commented on Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:
```diff
@@ -953,6 +966,319 @@ static ErrorOr<void> cascade_custom_properties(DOM::Element& element, Optional<C
+
+bool StyleComputer::Animation::is_done() const
+{
+    return progress.as_fraction() >= 0.9999f && iteration_count.has_value() && iteration_count.value() == 0;
```
there was a different place with three 9s previously — looks like a constant we could unify…? (some kind of `_threshold` or `_epsilon` as people tend to name it?)

alimpfard replied:
Unless it shows up as a broken animation, the actual number here doesn't really matter (you'd only observe this on say, `animation-duration: 10000s;` where 99.99% is 1 second, but even then, that error is fine imo for an initial impl, the "real" spec-compliant implementation of this function has other parameters anyway)

konradekk reviewed (commented):
minors, as always here… 😅

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/Parser/Token.h:
```diff
@@ -175,6 +175,22 @@ class Token {
         return token;
     }
 
+    static Token create_curly(Type curly_type)
+    {
+        VERIFY(curly_type == Type::OpenCurly || curly_type == Type::CloseCurly);
+        Token token;
+        token.m_type = curly_type;
+        return token;
+    }
+
+    static Token create_delim(String text)
+    {
+        Token token;
+        token.m_type = Type::Delim;
+        token.m_value = text;
+        return token;
+    }
+
```
These now aren't used anywhere.

AtkinsSJ commented on Userland/Libraries/LibWeb/CSS/StyleComputer.cpp:
```diff
@@ -1365,13 +1383,26 @@ ErrorOr<void> StyleComputer::compute_cascaded_values(StyleProperties& style, DOM
                         iteration_count = static_cast<size_t>(iteration_count_value->as_numeric().number());
 
+                    Animation::FillMode fill_mode { Animation::FillMode::None };
+                    if (auto fill_mode_value = style.maybe_null_property(PropertyID::AnimationFillMode); fill_mode_value && fill_mode_value->is_identifier()) {
+                        auto fill_mode_name = fill_mode_value->to_identifier();
+                        if (fill_mode_name == ValueID::Forwards)
+                            fill_mode = Animation::FillMode::Forward;
+                        if (fill_mode_name == ValueID::Backwards)
+                            fill_mode = Animation::FillMode::Backward;
+                        if (fill_mode_name == ValueID::None)
+                            fill_mode = Animation::FillMode::None;
+                        if (fill_mode_name == ValueID::Both)
+                            fill_mode = Animation::FillMode::Both;
+                    }
```
Adding the `animation-fill-mode` values to `Enums.json` would get you the to/from-identifier code for free.

AtkinsSJ reviewed (commented):
Just minor things. (Accidentally clicked the wrong button for the unused token-makers comment. :sweat_smile:)

These aren't blockers, just if you're happy to make the changes right away go for it, otherwise they can be made after merging.

Lubrsi commented on Userland/Libraries/LibWeb/CSS/StyleComputer.h:
```diff
@@ -126,17 +133,29 @@ class StyleComputer {
+    struct AnimationKeyFrameSet {
+        struct ResolvedKeyFrame {
+            struct UseInitial { };
+            Array<Variant<Empty, UseInitial, NonnullRefPtr<StyleValue const>>, to_underlying(last_property_id) + 1> resolved_properties;
```
This is uninitialized. Fixes https://github.com/SerenityOS/serenity/pull/19050#issuecomment-1564384089
```suggestion
            Array<Variant<Empty, UseInitial, NonnullRefPtr<StyleValue const>>, to_underlying(last_property_id) + 1> resolved_properties {};
```

alimpfard commented:
Changelog:
- Added `animation-direction` because I felt like it
- Fixed the uninitialised array issue @Lubrsi mentioned
- Switched to using Enums.json for `animation-{direction,fill-mode}` values
- Removed the unused Token::functions.

awesomekling commented:
Very cool! Let's bring it in :)
